backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt

```diff
@@ -0,0 +1,161 @@
+    configuration.getList(KonanConfigKeys.LIBRARIES_TO_COVER),
+    resolvedLibraries,
+    resolver,
+    FeaturedLibrariesReporter.forCoveredLibraries(configuration)
+)
+
+private sealed class FeaturedLibrariesReporter {
+
+    abstract fun reportIllegalKind(library: KonanLibrary)
+    abstract fun reportNotIncludedLibraries(includedLibraries: List<KonanLibrary>, remainingFeaturedLibraries: Set<File>)
+
+    object Silent: FeaturedLibrariesReporter() {
+        override fun reportIllegalKind(library: KonanLibrary) {}
+        override fun reportNotIncludedLibraries(includedLibraries: List<KonanLibrary>, remainingFeaturedLibraries: Set<File>) {}
+
+    }
+
+    abstract class BaseReporter(val configuration: CompilerConfiguration) : FeaturedLibrariesReporter() {
+        protected abstract fun illegalKindMessage(kind: String, libraryName: String): String
+        protected abstract fun notIncludedLibraryMessageTitle(): String
+
+        override fun reportIllegalKind(library: KonanLibrary) {
+            val kind = if (library.isInterop) "Interop" else "Default"
+            configuration.report(CompilerMessageSeverity.STRONG_WARNING, illegalKindMessage(kind, library.libraryName))
+        }
+
+        override fun reportNotIncludedLibraries(includedLibraries: List<KonanLibrary>, remainingFeaturedLibraries: Set<File>) {
+            val message = buildString {
+                appendln(notIncludedLibraryMessageTitle())
+                remainingFeaturedLibraries.forEach { appendln(it) }
+                appendln()
+                appendln("Included libraries:")
+                includedLibraries.forEach { appendln(it.libraryFile) }
+            }
+
+            configuration.report(CompilerMessageSeverity.STRONG_WARNING, message)
+        }
+    }
+
+    private class ExportedLibrariesReporter(configuration: CompilerConfiguration) : BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be exported with -Xexport-library"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are specified to be exported with -Xexport-library, but not included to the build:"
+    }
+
+    private class SourceLibrariesReporter(configuration: CompilerConfiguration) : BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be used as a source library"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are declared as source libraries with -Xsource-library, but not included to the build:"
+    }
+
+    private class CoveredLibraryReporter(configuration: CompilerConfiguration): BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be covered"
```

Comment:
Not sure if "covered" term can be used without mentioning "code coverage".

Reply:
I've used a construction from the corresponding help message. Looks like it should be fixed too.